File: cmd/pull.go
package cmd

import (
	"bufio"
	"errors"
	"fmt"
	"strings"
	"time"

	diagv1 "buf.build/gen/go/tldiagramcom/diagram/protocolbuffers/go/diag/v1"
	"connectrpc.com/connect"
	"github.com/mertcikla/tld-cli/client"
	"github.com/mertcikla/tld-cli/workspace"
	"github.com/spf13/cobra"
)

func newPullCmd(wdir *string) *cobra.Command {
	var force bool
	var dryRun bool

	c := &cobra.Command{
		Use:   "pull",
		Short: "Pull the current server state into local YAML files",
		Long: `Pull downloads the current diagram state from the server and overwrites
local YAML files. Use this after making changes in the frontend UI.

If you have local changes that haven't been applied yet, tld pull will warn
you before overwriting them. Use --force to skip the prompt.`,
		RunE: func(cmd *cobra.Command, _ []string) error {
			ws, err := loadWorkspaceWithHint(*wdir)
			if err != nil {
				return err
			}
			if err := ensureAPIKey(ws.Config.APIKey); err != nil {
				return err
			}

			targetOrg := ws.Config.OrgID
			if targetOrg == "" {
				return orgIDRequiredHint("org-id required in .tld.yaml")
			}

			lockFile, err := workspace.LoadLockFile(*wdir)
			if err != nil {
				return fmt.Errorf("load lock file: %w", err)
			}

			// Detect local changes if we have a previous hash to compare against
			if !force && lockFile != nil && lockFile.WorkspaceHash != "" {
				currentHash, err := workspace.CalculateWorkspaceHash(*wdir)
				if err != nil {
					return fmt.Errorf("calculate hash: %w", err)
				}
				if currentHash != lockFile.WorkspaceHash {
					fmt.Fprintf(cmd.OutOrStdout(), "Warning: local workspace has uncommitted changes.\n")
					fmt.Fprintf(cmd.OutOrStdout(), "Pull will overwrite them. Continue? [yes/no]: ")
					scanner := bufio.NewScanner(cmd.InOrStdin())
					if !scanner.Scan() {
						return errors.New("aborted")
					}
					answer := strings.TrimSpace(strings.ToLower(scanner.Text()))
					if answer != "yes" && answer != "y" {
						fmt.Fprintln(cmd.OutOrStdout(), "Pull cancelled.")
						return nil
					}
				}
			}

			c := client.New(ws.Config.ServerURL, ws.Config.APIKey, false)
			resp, err := c.ExportWorkspace(cmd.Context(), connect.NewRequest(&diagv1.ExportOrganizationRequest{
				OrgId: targetOrg,
			}))
			if err != nil {
				return withUnauthorizedHint("pull failed", err)
			}

			newWS := convertExportResponse(ws, resp.Msg)

			if dryRun {
				fmt.Fprintf(cmd.OutOrStdout(), "Would pull: %d elements, %d diagrams, %d connectors\n",
					len(newWS.Elements), countElementDiagrams(newWS), len(newWS.Connectors))
				return nil
			}

			// Perform surgical merge
			lastSyncMeta := &workspace.Meta{
				Elements:   make(map[string]*workspace.ResourceMetadata),
				Views:      make(map[string]*workspace.ResourceMetadata),
				Connectors: make(map[string]*workspace.ResourceMetadata),
			}
			if lockFile != nil && lockFile.Metadata != nil {
				lastSyncMeta = lockFile.Metadata
			}

			if force {
				if err := workspace.Save(newWS); err != nil {
					return fmt.Errorf("force save workspace: %w", err)
				}
			} else {
				if err := workspace.MergeWorkspace(*wdir, newWS, lastSyncMeta, ws.Meta); err != nil {
					return fmt.Errorf("merge workspace: %w", err)
				}
			}

			hash, err := workspace.CalculateWorkspaceHash(*wdir)
			if err != nil {
				return fmt.Errorf("calculate hash: %w", err)
			}
			if lockFile == nil {
				lockFile = &workspace.LockFile{Version: "v1"}
			}
			versionID := fmt.Sprintf("pull-%s", time.Now().UTC().Format(time.RFC3339))
			workspace.UpdateLockFile(lockFile, versionID, "pull", &workspace.ResourceCounts{
				Elements:   len(newWS.Elements),
				Views:      countElementDiagrams(newWS),
				Connectors: len(newWS.Connectors),
			}, hash, nil, newWS.Meta)
			lockFile.Resources.Elements = len(newWS.Elements)
			lockFile.Resources.Views = countElementDiagrams(newWS)
			lockFile.Resources.Connectors = len(newWS.Connectors)
			if err := workspace.WriteLockFile(*wdir, lockFile); err != nil {
				return fmt.Errorf("write lock file: %w", err)
			}

			fmt.Fprintf(cmd.OutOrStdout(), "Pulled %d elements, %d diagrams, %d connectors\n",
				len(newWS.Elements), countElementDiagrams(newWS), len(newWS.Connectors))

			return nil
		},
	}

	c.Flags().BoolVar(&force, "force", false, "overwrite local changes without prompting")
	c.Flags().BoolVar(&dryRun, "dry-run", false, "show what would be pulled without writing")
	return c
}

File: src/cli/pull.rs
use crate::client;
use crate::client::diagv1;
use crate::error::TldError;
use crate::output;
use crate::workspace;
use clap::Args;
use tonic::Request;

#[derive(Args, Debug, Clone)]
pub struct PullArgs {
    /// Apply without prompting for confirmation
    #[arg(long, default_value = "false")]
    pub force: bool,
    /// show what would be pulled without writing
    #[arg(long, default_value = "false")]
    pub dry_run: bool,
}

pub async fn exec(args: PullArgs, wdir: String) -> Result<(), TldError> {
    let ws = workspace::load(&wdir)?;

    if ws.config.org_id.is_empty() {
        return Err(TldError::Generic(
            "Org ID required in .tld.yaml".to_string(),
        ));
    }

    if ws.config.api_key.is_empty() {
        return Err(TldError::Generic(
            "No API key found. Run 'tld login' first.".to_string(),
        ));
    }

    output::print_info("Pulling latest state from server...");

    let mut client =
        client::new_workspace_client(&ws.config.server_url, &ws.config.api_key).await?;

    let req = Request::new(diagv1::ExportOrganizationRequest {
        org_id: ws.config.org_id.clone(),
        api_key: None,
    });

    let spinner = output::new_spinner("Contacting server...");
    let resp = client.export_workspace(req).await?.into_inner();
    spinner.finish_and_clear();

    let server_ws = workspace::conversion::from_export_response(
        &wdir,
        ws.config.clone(),
        ws.workspace_config.clone(),
        ws.meta.as_ref(),
        resp,
    );

    // Load last sync state from lockfile for 3-way merge
    let lock_file = workspace::load_lock_file(&wdir)?.unwrap_or_default();
    let last_sync_meta = lock_file.to_meta();
    let current_meta = ws.meta.clone().unwrap_or_default();

    if args.dry_run {
        output::print_info(&format!(
            "Would pull {} elements and {} connectors.",
            server_ws.elements.len(),
            server_ws.connectors.len()
        ));
        return Ok(());
    }

    if args.force {
        workspace::save(&server_ws)?;
        output::print_ok("Workspace updated (forced overwrite).");
    } else {
        workspace::merger::merge_workspace(&wdir, &server_ws, &last_sync_meta, &current_meta)?;
        output::print_ok("Workspace merged with server state.");
    }

    // Update lockfile
    let lf = workspace::lockfile::LockFile::from_workspace(&server_ws);
    workspace::lockfile::save_lock_file(&wdir, &lf)?;

    Ok(())
}
